### PR TITLE
Include hook's object name when profiling

### PR DESCRIPTION
**Pull request by daniellepintz**
## What does this PR do?
When profiling, we should distinguish between a LM hook vs callback hook vs TTP hook. 
Part of #8506

## Before submitting

- [x] Was this **discussed/approved** via a GitHub issue? (not for typos and docs)
- [x] Did you read the [contributor guideline](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/.github/CONTRIBUTING.md), **Pull Request** section?
- [x] Did you make sure your **PR does only one thing**, instead of bundling different changes together?
- [x] Did you make sure to **update the documentation** with your changes? (if necessary)
- [x] Did you write any **new necessary tests**? (not for typos and docs)
- [x] Did you verify new and **existing tests pass** locally with your changes?
- [x] Did you list all the **breaking changes** introduced by this pull request?
- [x] Did you **update the [CHANGELOG](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/CHANGELOG.md)**? (not for typos, docs, test updates, or internal minor changes/refactorings)

## PR review

Anyone in the community is welcome to review the PR.
Before you start reviewing make sure you have read [Review guidelines](https://github.com/PyTorchLightning/pytorch-lightning/wiki/Review-guidelines). In short, see the following bullet-list:

- [x] Is this pull request ready for review? (if not, please submit in draft mode)
- [x] Check that all items from **Before submitting** are resolved
- [x] Make sure the title is self-explanatory and the description concisely explains the PR
- [x] Add labels and milestones (and optionally projects) to the PR so it can be classified

## Did you have fun?

Make sure you had fun coding 🙃

### CODE REVIEW

File: pytorch_lightning/trainer/trainer.py [hunk 1]
```diff
@@ -1562,7 +1561,7 @@ def _call_callback_hooks(
             for callback in self.callbacks:
                 fn = getattr(callback, hook_name)
                 if callable(fn):
-                    with self.profiler.profile(hook_name):
+                    with self.profiler.profile(f"{callback.__class__.__name__}.{hook_name}"):
```
Comment (tchaton) [inline pytorch_lightning/trainer/trainer.py:1564]: How would you differentiate when there are several callbacks of the same type?

File: pytorch_lightning/trainer/trainer.py [hunk 2]
```diff
@@ -1520,8 +1520,7 @@ def _call_lightning_module_hook(
         prev_fx_name = pl_module._current_fx_name
         pl_module._current_fx_name = hook_name
 
-        # TODO: when profiling separate hook name by hook object name (e.g. Callback, LM)
-        with self.profiler.profile(hook_name):
+        with self.profiler.profile(f"{pl_module.__class__.__name__}.{hook_name}"):
```
Comment (ananthsub) [inline pytorch_lightning/trainer/trainer.py:1523]: Do we need the prefix to be deterministic? Right now if we have `MyCustomLightningModule.{hook_name}` it can be hard for the profiler to distinguish this hook vs `MyCustomPlugin.{hook_name}` 

Something like `LightningModule.MyCustomLightningModule.{hook_name}` could provide a descriptive record and allow downstream systems to better process this data, at the cost of verbosity

File: pytorch_lightning/trainer/trainer.py [hunk 3]
```diff
@@ -1560,7 +1559,7 @@ def _call_callback_hooks(
             for callback in self.callbacks:
                 fn = getattr(callback, hook_name)
                 if callable(fn):
-                    with self.profiler.profile(hook_name):
+                    with self.profiler.profile(f"{callback.__class__.__name__}.{callback.state_key}.{hook_name}"):
```
Comment (ananthsub) [inline pytorch_lightning/trainer/trainer.py:1562]: same here -`Callback.{state_key}.{hook_name}` ?

also `callback.__class__.__name__` could be redundant w/ state_key. 

State key should include the qualname already. 
https://github.com/PyTorchLightning/pytorch-lightning/blob/5576fbc5f9a7d0bc71ad26b8b54775110e675808/pytorch_lightning/callbacks/base.py#L36-L58


Here's an example from the early stopping callback: https://github.com/PyTorchLightning/pytorch-lightning/blob/5576fbc5f9a7d0bc71ad26b8b54775110e675808/pytorch_lightning/callbacks/early_stopping.py#L124-L126

**Review by ananthsub — COMMENTED** [review]
This is really great to have more granular profiling available for debugging!

File: pytorch_lightning/trainer/trainer.py (hunk 2 above)
Comment (daniellepintz) [reply pytorch_lightning/trainer/trainer.py:1523]: Makes sense! will update

File: tests/profiler/test_profiler.py [hunk 4]
```diff
@@ -311,9 +311,10 @@ def test_pytorch_profiler_trainer_ddp(tmpdir, pytorch_profiler):
     )
     trainer.fit(model)
 
-    expected = {"BoringModel.validation_step"}
+    expected = {"LightningModel.BoringModel.validation_step"}
```
Comment (ananthsub) [inline tests/profiler/test_profiler.py:314]:
```suggestion
    expected = {"LightningModule.BoringModel.validation_step"}
```

**Review by ananthsub — APPROVED** [review]

**Review by rohitgr7 — APPROVED** [review]

File: tests/profiler/test_profiler.py [hunk 5]
```diff
@@ -552,3 +553,25 @@ def test_pytorch_profiler_raises_warning_for_limited_steps(tmpdir, trainer_confi
         getattr(trainer, trainer_fn)(model)
         assert trainer.profiler._schedule is None
         warning_cache.clear()
+
+
+def test_profile_callbacks(tmpdir):
+    """Checks is profiling callbacks works correctly, specifically when there are two of the same callback type."""
```
Comment (awaelchli) [inline tests/profiler/test_profiler.py:559]:
```suggestion
    """Checks if profiling callbacks works correctly, specifically when there are two of the same callback type."""
```

**Review by awaelchli — COMMENTED** [review]
sorry for nit, isn't it better to mention this in changelog?

File: tests/profiler/test_profiler.py [hunk 6]
```diff
@@ -343,16 +348,16 @@ def test_pytorch_profiler_trainer_fit(fast_dev_run, boring_model_cls, tmpdir):
     trainer = Trainer(default_root_dir=tmpdir, max_epochs=1, fast_dev_run=fast_dev_run, profiler=pytorch_profiler)
     trainer.fit(model)
 
-    assert sum(e.name == "validation_step" for e in pytorch_profiler.function_events)
+    assert sum(
+        e.name == "TrainingTypePlugin.SingleDevicePlugin.validation_step" for e in pytorch_profiler.function_events
```
Comment (awaelchli) [inline tests/profiler/test_profiler.py:352]: what is the motivation behind this pattern of including the base class? Wouldn't it have to be separated by a different symbol?

**Review by awaelchli — APPROVED** [review]

File: tests/profiler/test_profiler.py [hunk 7]
```diff
@@ -311,9 +311,14 @@ def test_pytorch_profiler_trainer_ddp(tmpdir, pytorch_profiler):
...
-    expected = {"validation_step"}
+    expected = {"LightningModule.BoringModel.validation_step"}
     if not _KINETO_AVAILABLE:
-        expected |= {"training_step_and_backward", "training_step", "backward"}
+        expected |= {
+            "training_step_and_backward",
+            "LightningModule.BoringModel.training_step",
+            "LightningModule.BoringModel.backward",
+        }
+    print([e.name for e in pytorch_profiler.function_events])
```
Comment (carmocca) [inline tests/profiler/test_profiler.py:321]: debugging statement
```suggestion
```

File: tests/profiler/test_profiler.py (hunk 6 above)
Comment (carmocca) [reply tests/profiler/test_profiler.py:352]: I also find it confusing as it's not something that exists. I suggest removing the base class

File: tests/profiler/test_profiler.py [hunk 8]
```diff
@@ -343,16 +348,16 @@ def test_pytorch_profiler_trainer_fit(fast_dev_run, boring_model_cls, tmpdir):
...
+    )
 
     path = pytorch_profiler.dirpath / f"fit-{pytorch_profiler.filename}.txt"
     assert path.read_text("utf-8")
 
     if _KINETO_AVAILABLE:
         files = sorted(file for file in os.listdir(tmpdir) if file.endswith(".json"))
         assert any(f"fit-{pytorch_profiler.filename}" in f for f in files)
-        path = pytorch_profiler.dirpath / f"fit-{pytorch_profiler.filename}.txt"
-        assert path.read_text("utf-8")
```
Comment (carmocca) [inline tests/profiler/test_profiler.py:355]: Why is this gone?

File: tests/profiler/test_profiler.py (hunk 6 above)
Comment (daniellepintz) [reply tests/profiler/test_profiler.py:352]: see @ananthsub's comments above

File: tests/profiler/test_profiler.py (hunk 7 above)
Comment (daniellepintz) [reply tests/profiler/test_profiler.py:321]: Will remove once I can get this test to pass :/ (https://dev.azure.com/PytorchLightning/pytorch-lightning/_build/results?buildId=47638&view=logs&j=3afc50db-e620-5b81-6016-870a6976ad29&t=d9f671c5-a304-5675-5394-961fd7f98b9b)

File: tests/profiler/test_profiler.py (hunk 8 above)
Comment (daniellepintz) [reply tests/profiler/test_profiler.py:355]: it seems to me duplicated lines of code? see 348-349

**Comment by daniellepintz** [comment]
@awaelchli mention what in changelog? I do have an entry in the changelog

File: tests/profiler/test_profiler.py (hunk 6 above)
Comment (ananthsub) [reply tests/profiler/test_profiler.py:352]: @awaelchli @carmocca this would be useful for automatically categorizing profiler data, especially for determining what came from the module vs callback vs strategy

Comment (carmocca) [reply tests/profiler/test_profiler.py:352]: I understand that, but the current structure makes it sound like `TrainingTypePlugin.SingleDevicePlugin.validation_step` is an attribute that exists but that's correct in Python.

I see two options:
- Drop the base class
- Remove the ambiguity with different notation, for example `[TrainingTypePlugin] SingleDevicePlugin.validation_step`

Comment (awaelchli) [reply tests/profiler/test_profiler.py:352]: notation with brackets sounds good to me

**Comment by awaelchli** [comment]
I meant an entry under "changed" that the profiler now indexes and displays the names of the hooks with a new pattern (base class).(class).(hook name)

File: tests/profiler/test_profiler.py (hunk 6 above)
Comment (daniellepintz) [reply tests/profiler/test_profiler.py:352]: Sg, will update!

**Comment by daniellepintz** [comment]
@awaelchli sure, I already have an entry under changed, but I can edit it

**Review by Borda — APPROVED** [review]

**Review by carmocca — DISMISSED** [review]
Requesting changes so I can review before merge :), ping me when you update it!

**Comment by daniellepintz** [comment]
I'm really stumped on this test failure https://dev.azure.com/PytorchLightning/pytorch-lightning/_build/results?buildId=47638&view=logs&j=3afc50db-e620-5b81-6016-870a6976ad29&t=d9f671c5-a304-5675-5394-961fd7f98b9b

Could anyone who has worked with profilers before try taking a look? cc @kaushikb11 @carmocca @rohitgr7

File: pytorch_lightning/profiler/pytorch.py [hunk 9]
```diff
@@ -340,11 +341,11 @@ def _total_steps(self) -> int:
...
         if self._schedule.is_training:
             return trainer.num_training_batches
-        if self._schedule._current_action == "validation_step":
+        if self._schedule._current_action.endswith("validation_step"):
             return sum(trainer.num_val_batches) + sum(trainer.num_sanity_val_batches)
-        if self._schedule._current_action == "test_step":
+        if self._schedule._current_action.endswith("test_step"):
             return sum(trainer.num_test_batches)
-        if self._schedule._current_action == "predict_step":
+        if self._schedule._current_action.endswith("predict_step"):
```
Comment (rohitgr7) [inline pytorch_lightning/profiler/pytorch.py:348]: these changes are required at other places in the same file. Specifically `has_finished`, `_step` and `num_step`. make sure to check if self._current_action is None before using `endswith` there.

File: tests/profiler/test_profiler.py [hunk 10]
```diff
@@ -310,10 +310,14 @@ def test_pytorch_profiler_trainer_ddp(tmpdir, pytorch_profiler):
...
     if not _KINETO_AVAILABLE:
-        expected |= {"training_step_and_backward", "training_step", "backward"}
+        expected |= {
+            "training_step_and_backward",
+            "[TrainingTypePlugin]DDPPlugin.training_step",
+            "[TrainingTypePlugin]DDPPlugin.backward",
+        }
+    print([e.name for e in pytorch_profiler.function_events])
     for name in expected:
         assert sum(e.name == name for e in pytorch_profiler.function_events), name
```
Comment (rohitgr7) [inline tests/profiler/test_profiler.py:323]: line 337:
```py
assert any(f"{local_rank}-validation_step" in f for f in files)
```
to
```py
assert any(f"{local_rank}-[TrainingTypePlugin]DDPPlugin.validation_step" in f for f in files)
```

File: pytorch_lightning/profiler/pytorch.py (hunk 9 above)
Comment (daniellepintz) [reply pytorch_lightning/profiler/pytorch.py:348]: completely missed this, thanks so much!

**Review by ananthsub — COMMENTED** [review]
heads up, there will be conflicts between this and #11142 since the naming is changing
cc @four4fish

**Comment by daniellepintz** [comment]
@carmocca ready for your review! All tests passing

File: pytorch_lightning/trainer/trainer.py [hunk 11]
```diff
@@ -1585,7 +1584,7 @@ def _call_ttp_hook(
         if not callable(fn):
             return
 
-        with self.profiler.profile(hook_name):
+        with self.profiler.profile(f"[TrainingTypePlugin]{self.training_type_plugin.__class__.__name__}.{hook_name}"):
```
Comment (carmocca) [inline pytorch_lightning/trainer/trainer.py:1587]: Use `Strategy` already insteaad of `TrainingTypePlugin`. Avoids follow-up changes.

File: pytorch_lightning/profiler/pytorch.py [hunk 12]
```diff
@@ -132,44 +132,51 @@ def reset(self):
...
         return self._current_action is not None and (
-            self._current_action.startswith("optimizer_step_with_closure_") or self._current_action == "training_step"
+            self._current_action.startswith("optimizer_step_with_closure_")
+            or self._current_action.endswith("training_step")
         )
 
     @property
     def num_step(self) -> int:
+        if self._current_action is None:
+            return 0
```
Comment (carmocca) [inline pytorch_lightning/profiler/pytorch.py:142]: I believe we could have all these `None` checks in one single place instead: `__call__`

```diff
diff --git a/pytorch_lightning/profiler/pytorch.py b/pytorch_lightning/profiler/pytorch.py
index c697fcad5..580bb57ff 100644
--- a/pytorch_lightning/profiler/pytorch.py
+++ b/pytorch_lightning/profiler/pytorch.py
@@ -131,15 +131,12 @@ class ScheduleWrapper:
 
     @property
     def is_training(self) -> bool:
-        return self._current_action is not None and (
-            self._current_action.startswith("optimizer_step_with_closure_")
-            or self._current_action.endswith("training_step")
+        return self._current_action.startswith("optimizer_step_with_closure_") or self._current_action.endswith(
+            "training_step"
         )
 
     @property
     def num_step(self) -> int:
-        if self._current_action is None:
-            return 0
         if self.is_training:
             return self._num_optimizer_step_with_closure
         if self._current_action.endswith("validation_step"):
@@ -151,8 +148,6 @@ class ScheduleWrapper:
         return 0
 
     def _step(self) -> None:
-        if self._current_action is None:
-            return
         if self.is_training:
             self._num_optimizer_step_with_closure += 1
         elif self._current_action.endswith("validation_step"):
@@ -168,8 +163,6 @@ class ScheduleWrapper:
 
     @property
     def has_finished(self) -> bool:
-        if self._current_action is None:
-            return False
         if self.is_training:
             return self._optimizer_step_with_closure_reached_end
         if self._current_action.endswith("validation_step"):
@@ -182,12 +175,12 @@ class ScheduleWrapper:
 
     def __call__(self, num_step: int) -> "ProfilerAction":
         # ignore the provided input. Keep internal state instead.
-        if self.has_finished:
+        if self._current_action is None or self.has_finished:
             return ProfilerAction.NONE
 
         self._step()
         action = self._schedule(max(self.num_step, 0))
-        if action == ProfilerAction.RECORD_AND_SAVE and self._current_action is not None:
+        if action == ProfilerAction.RECORD_AND_SAVE:
             if self.is_training:
                 self._optimizer_step_with_closure_reached_end = True
             elif self._current_action.endswith("validation_step"):
```

This seems to work locally.

Comment (daniellepintz) [reply pytorch_lightning/profiler/pytorch.py:142]: Much better! Thanks

**Comment by daniellepintz** [comment]
Ready to be merged!
